Add M1 support to build

Stonedestroyer opened the pull request:
Depends on #88, thanks @Walkyst for your work there! 

This adds M1 as a target compilation from x86_x64 mac runner, in addition min deployment target with the default XCode supplied is 10.12.6, as Lavalink will only run on 10.12, re @aikaterna.

Thanks to @aikaterna and @jack1142 for assisting with this PR. 

Some improvements, we could ditch Java 15 and use the built-in Github runner java versions https://github.com/actions/runner-images/blob/main/images/macos/macos-11-Readme.md#java but it's not really needed, unless you want me to do that @natanbc 

TODO:

- [x] Solve the native folder for darwin creating several files

Walkyst committed:
Fixed mac build

Stonedestroyer committed:
Add m1

Stonedestroyer committed:
Remove avx2 after combining

Stonedestroyer committed:
Sets min deployment target to 10.21

Lowest version that Lavalink will run under.

Co-authored-by: aikaterna <20862007+aikaterna@users.noreply.github.com>
Co-authored-by: Jakub Kuczys <6032823+jack1142@users.noreply.github.com>

Stonedestroyer committed:
Don't remove avx2 file

natanbc commented on natives/toolchains.gradle:
```diff
@@ -12,7 +13,26 @@ task buildWin32 {}
+    doLast {
+        def stdout = new ByteArrayOutputStream()
+        exec {
+            commandLine 'lipo', avx2FilePath.getAbsoluteFile(), aarch64FilePath.getAbsoluteFile(), '-output', finalFilePath.getAbsoluteFile(), '-create'
```
Should be `x86FilePath` instead of `avx2FilePath`, since on systems without avx2 this file would be loaded as a fallback

Stonedestroyer replied:
Thank you

Stonedestroyer committed:
Bundle x86 instead of avx2 in lipo

natanbc committed:
Build arm64 darwin natives as well

natanbc committed:
Split arm64 darwin target argument

`withArguments` deals with already split arguments, so arguments with spaces are passed as a single argument

natanbc committed:
Fix both x86-64 and arm64 darwin targets writing to the same output file

natanbc committed:
Oops

natanbc committed:
println debugging here we go

natanbc committed:
Update the right variable, don't add architecture to avx2 variant

natanbc committed:
Maybe the linker needs a target too

natanbc committed:
Cleanup pt1 (exec defaults to stdout, unused avx2 variable)

natanbc committed:
Cleanup pt2 (remove debug prints)

Stonedestroyer commented:
This is blocked by the native lib loader not supporting mac m1, 

Per tests by @aikaterna loading natives on mac m1 would load the avx2 one, as aarch64 does not contain support for avx2 it will crash. 

The issue to my understanding the upstream library cpu_features for native lib detector does not support m1. The support is pending on the upstream library, see https://github.com/natanbc/native-loader/issues/1. 

A solution would be to do something smilar to , https://github.com/aikaterna/lavadsp/blob/f6fce3c9d3c630a77061780f2902c551416c5a12/src/main/java/com/github/natanbc/lavadsp/natives/TimescaleNativeLibLoader.java#L35. 
This would ensure we do not load avx2 on Mac m1 arm, and would work until cpu_features gets updated or we get a better solution. 

The solution above is tested and works.

natanbc commented:
Started working on a rewrite of native-loader, which will also support M1 https://github.com/natanbc/native-loader/tree/native-rewrite
